### DISCUSSION TITLE
Documentation for the addition of event based messaging.

### DISCUSSION THREAD

**Discussion by mkmer** [fronzbot/blinkpy#770] [Q&A]
Goal here to start documenting the event based message system used by Blinkapp in android.

- It appears the system uses FCM (firebase messaging) to send event based messages.

- Adding the following key to the login json message appears to link the messaging system to the app: 
     `"notification_key": "<redacted key here>",`

- An official login JSON message sent from android (blinkpy only fills a subset):
`{
  "app_version": "6.29.0 (28173197) #28f4ca29a",
  "client_name": "Pixel 3",
  "client_type": "android",
  "device_identifier": "Google Pixel 3,google/blueline",
  "email": "<redacted>",
  "notification_key": "<redacted>",
  "os_version": "12",
  "password": "<redacted>",
  "reauth": true,
  "unique_id": "<redacted>"
}`

- After a notification message is received by android the app replies with:

`METHOD: POST
URL
https://rest-u034.immedia-semi.com/api/v2/notification`
`HEADERS
Accept-Encoding:
gzip
APP-BUILD:
ANDROID_28173197
Connection:
Keep-Alive
Content-Length:
57
Content-Type:
application/json; charset=UTF-8
Host:
rest-u034.immedia-semi.com
LOCALE:
en_US
User-Agent:
6.29.0ANDROID_28173197
X-Blink-Time-Zone:
America/Detroit`

`{
  "status": "received",
  "transaction_id": "oMFXIR25gyZy9GyA"
}`
Transaction_id value must be sent via the FCM message.

- The `notification_key` appears to be provided when the application connects to the FCM server.  So we will first need to connect to FCM, get the key, then login to the app providing the given key.

- There is a library (that needs to be ported to aiohttp): https://github.com/Francesco149/push_receiver

- At this point, I have tried subscribing and using the key in `blinkpy`, but have not been able to receive notifications via the `push_reciever` app.  I suspect it needs the `sender_id` configured, but can not located the value at this point.

Helpers wanted :)

**Comment by gwyntel-git** [comment]
Hey there, not sure if you're still interested in help with this, but the sender_id appears to be 351117235468
